tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,318 @@
+            conn->actual_protocol_version = S2N_TLS13;
+            
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&out, CHOSEN_PSK_WIRE_INDEX_SIZE));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&out, chosen_psk_wire_index));
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_server_psk_extension.recv(conn, &out), S2N_ERR_INVALID_ARGUMENT);
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+
+            for (size_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk));
+                EXPECT_NOT_NULL(psk);
+            }
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+        /* Receive valid server preshared extension recv */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+
+            for (size_t i = 0; i < s2n_array_len(test_cases); i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+                EXPECT_SUCCESS(s2n_psk_init(psk, S2N_PSK_TYPE_EXTERNAL));
+                EXPECT_SUCCESS(s2n_psk_new_identity(psk, test_cases[i].identity, test_cases[i].identity_size));
+                psk->hmac_alg = test_cases[i].hmac_alg;
+            }
+
+            uint8_t chosen_psk_wire_index = 0;
+            conn->actual_protocol_version = S2N_TLS13;
+
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&out, CHOSEN_PSK_WIRE_INDEX_SIZE));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&out, chosen_psk_wire_index));
+
+            EXPECT_SUCCESS(s2n_server_psk_extension.recv(conn, &out));
+            EXPECT_NOT_NULL(conn->psk_params.chosen_psk);
+
+            /* Validate that the PSKs not chosen are wiped */
+            for (size_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk));
+                if(i == conn->psk_params.chosen_psk_wire_index) {
+                    EXPECT_NOT_NULL(psk);
+                } else {
+                    EXPECT_EQUAL(psk->identity.data, NULL);
+                    EXPECT_EQUAL(psk->identity.size, 0);
+                    EXPECT_EQUAL(psk->secret.data, NULL);
+                    EXPECT_EQUAL(psk->secret.size, 0);
+                    EXPECT_EQUAL(psk->early_secret.data, NULL);
+                    EXPECT_EQUAL(psk->early_secret.size, 0);
+                }
+            }
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+    }
+    
+    /* Functional test */
```

Comment:
A full functional test would probably send and receive the ServerHello, which would have caught that you hadn't added the extension to the ServerHello list.